### ISSUE TITLE
AMP documentation on command errors is reversed or code is

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2273]
|<img alt="htj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| htj reported|
|-|-|
|Trac ID|trac#2273|
|Type|defect|
|Created|2006-12-06 14:04:27Z|

line 1038 in protocols/amp.py says:

    @cvar errors: A mapping of error tag to exception type.

which is wrong because errors have to be mapped like this to work:

    errors = {SomeException: 'SOME_ERROR'}

The easy solution is to flip the text in the docstring so it reads:

    @cvar errors: A mapping of exceptions types to error tags.

However, I think that what the docstring says is better. By having ERROR_TAG -> Exception, it is possible to map several error tags to the same exception. This is not possible with the current scheme, as an exception maps to an error tag. Hence one exception cannot map to several tags, unless we start using lists, but that just makes it more nasty.

I know this is an API change, but since AMP has not yet been in official release it should be doable.

<details><summary>Searchable metadata</summary>

```
trac-id__2273 2273
type__defect defect
reporter__htj htj
priority__normal normal
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1165413867000000 1165413867000000
changetime__1165686227000000 1165686227000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @radix|
|-|-|

You should do something about this, either changing the docs or the code.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix set owner to @mithrandi|
|-|-|

Ready for review in fix-amp-docstring-2273


For the reviewer's convenience, I include the entire changeset here:


```
-    @cvar errors: A mapping of error tag to exception type.
+    @cvar errors: A mapping of exception type to error tag.
```

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi set owner to @radix|
|-|-|

As best I can tell, this change is correct. Go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix set status to `closed`|
|-|-|

(In [18850]) Merge fix-amp-docstring-2273

Author: radix
Reviewer: mithrandi
Fixes [#2273](https://github.com/twisted/twisted/issues/2273)

One is meant to pass a mapping of exception types to tags
as the 'errors' argument to Command, not a mapping of tags to exception
types.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
